runtime/ts/storage/pouchdb/pouch-db-collection.ts

```diff
@@ -0,0 +1,364 @@
+   * Triggered when the storage key has been modified.  For now we
+   * just refetch and trigger listeners.  This is fast since the data
+   * is synced locally.
+   */
+  public onRemoteStateSynced() {
+    // updates internal state
+    this.getModel();
+  }
+
+  /**
+   * Updates the local model cache from PouchDB and returns the CRDT
+   * model for use.
+   */
+  private async getModel(): Promise<CrdtCollectionModel> {
+    try {
+      const result = await this.db.get(this.pouchDbKey.location);
+      
+      // compare revisions
+      if (this._rev !== result._rev) {
+        // remote revision is different, update local copy.
+        this._model = new CrdtCollectionModel(result['model']);
+        this._rev = result._rev;
+        this.version++;  // yuck.
+        // TODO(lindner): fire change events here?
+      }
+
+    } catch (err) {
+      this._model = new CrdtCollectionModel();
+      this._rev = undefined;
+      // TODO(lindner): throw
+    }
+    return this._model;
+  }
+
+
+  /**
+   * Provides a way to apply changes to the model in a way that will result in the
+   * crdt being written to the underlying PouchDB.
+   * 
+   * - A new entry is stored if it doesn't exists.
+   * - If the existing entry is available it is fetched and the
+   *   internal state is updated.
+   * - A copy of the CRDT model is passed to the modelMutator, which may change it.
+   * - If the model is mutated by `modelMutator`, write a new revision and update the local
+   *   cached copy.
+   *
+   * @param modelMutator allows for modifying a copy of the underlying crdt model.
+   */
+  private async getModelAndUpdate(modelMutator: CrdtCollectionModelMutator): Promise<CrdtCollectionModel> {
+    // Keep retrying the operation until it succeeds.
+    while (1) {
+      let doc;
+      //: PouchDB.Core.IdMeta & PouchDB.Core.GetMeta & Model & {referenceMode: boolean, type: {}};
+
+      let notFound = false;
+      try {
+        doc = await this.db.get(this.pouchDbKey.location);
+        // as PouchDB.Core.IdMeta & PouchDB.Core.GetMeta & Model & {referenceMode: boolean, type: };
+
+        // Check remote doc.
+        // TODO(lindner): refactor with getModel above.
+        if (this._rev !== doc._rev) {
+          // remote revision is different, update local copy.
+          this._model = new CrdtCollectionModel(doc['model']);
+          this._rev = doc._rev;
+          this.version++;  // yuck.
+          // TODO(lindner): fire change events here?
+        }
+
+      } catch (err) {
+        if (err.name !== 'not_found') {
+          throw err;
+        }
+        notFound = true;
+        // setup basic doc, model/version updated below.
+        doc = {
+          _id: this.pouchDbKey.location,
+          referenceMode: this.referenceMode,
+          type: this.type.toLiteral()
+        };
+      }
+
+      // Run the mutator on a copy of the existing model
+      const newModel = modelMutator(new CrdtCollectionModel(this._model.toLiteral()));
```

Comment:
We call `toLiteral` for `_model` and `newModel` multiple times here and below, I think it's possible it could be expensive, not sure if we need new instances or if we could call once and reuse.

Reply:
added TODO